src/Features/Core/Portable/EmbeddedLanguages/Classification/AbstractEmbeddedLanguageClassificationService.cs

```diff
@@ -83,22 +80,29 @@ internal abstract class AbstractEmbeddedLanguageClassificationService :
                 _cancellationToken = cancellationToken;
             }
 
-            public void Recurse(SyntaxNode node)
+            public void VisitTokens(SyntaxNode node)
             {
-                _cancellationToken.ThrowIfCancellationRequested();
-                if (node.Span.IntersectsWith(_textSpan))
+                using var pooledStack = SharedPools.Default<Stack<SyntaxNodeOrToken>>().GetPooledObject();
+                var stack = pooledStack.Object;
+                stack.Push(node);
+                while (!stack.IsEmpty())
                 {
-                    foreach (var child in node.ChildNodesAndTokens())
+                    _cancellationToken.ThrowIfCancellationRequested();
+                    var currentNodeOrToken = stack.Pop();
+                    if (currentNodeOrToken.IsNode)
                     {
-                        if (child.IsNode)
-                        {
-                            Recurse(child.AsNode()!);
-                        }
-                        else
+                        if (currentNodeOrToken.Span.IntersectsWith(_textSpan))
                         {
-                            ProcessToken(child.AsToken());
+                            foreach (var child in currentNodeOrToken.ChildNodesAndTokens().Reverse())
+                            {
+                                stack.Push(child);
+                            }
                         }
                     }
+                    else
+                    {
+                        ProcessToken(currentNodeOrToken.AsToken());
```

Comment:
is it ok that we're processing all tokens, even those that doesn't intersect with _TextSpan?

Reply:
I think all the token's parents are intersects with the _textspan, which should be enough?
But I added it to the if check to make sure